### PR TITLE
TYP: Fix core/groupby/generic.py type ignores

### PR DESCRIPTION
**Pull request by FactorizeD**
xref #37715

Two things to mention:

- not a huge fan of casts but couldn't find another way to fix the issues;
- added `Index()` in line 277 because `columns` attribute of a dataframe is supposed to be of Index type.

### CODE REVIEW

File: pandas/core/groupby/generic.py
Comment (twoertwein) [inline pandas/core/groupby/generic.py:278]: Could this change the runtime behavior/affect performance? @jbrockmendel 

If `.columns` is a setter of a property (not sure whether this is the case), it would do any conversions internally: if that is the case, it might be worth to double-check the signature of the setter.

File: pandas/core/groupby/generic.py
Comment (twoertwein) [inline pandas/core/groupby/generic.py:629]: If possible, it would be great to instead add overloads to [factorize](https://github.com/pandas-dev/pandas/blob/1b3bb88008011f8ce7e3ec65fcc0c31fb838352b/pandas/core/algorithms.py#L576).

File: pandas/core/groupby/generic.py
Comment (twoertwein) [inline pandas/core/groupby/generic.py:648]: @simonjayhawkins might be a better person to judge when/where to use casts. I think I also sometime use them where I shouldn't use them :)

File: pandas/core/groupby/generic.py
Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:278]: yep. let's not change code in a typing PR. Have opened #46565 to address setter issue.

Comment (FactorizeD) [reply pandas/core/groupby/generic.py:278]: Right, didn't fully think of the potential consequences, sorry about that. Will remember not to change the code in the future typing PRs. I will wait for #46565 to be merged and then will continue working on this one

Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:278]: > Right, didn't fully think of the potential consequences, sorry about that.

IIRC we have had regressions where index labels are lost and regressions where regular Indexes are created where the code in ensure_index would have created a MultiIndex. Not sure if that is an issue here, but normally better to keep code changes out of typing PRs.

Having said that, where we are removing ignores, it could be that mypy is finding a latent bug or that the code could be refactored to facilitate static type checking (and in these cases normally improve readability of the code for humans too)

> I will wait for #46565 to be merged and then will continue working on this one

no need to wait, typing PRs regularly get bogged down in discussion. I would just remove this change for now and continue with the rest of it (will try to get to the other questions on this PR soon)

File: pandas/core/groupby/generic.py
Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:629]: the revealed type of `val` is `Any`. `val = self.obj._values`, The revealed type of `self.obj` is correctly `Series`, but the `_value` property of Series is untyped.

So I don't think overloading factorize would help here till `val` does not resolve to `Any`

I don't think in general should replace ignores with casts. We should be careful that we are not lying to mypy. If it is clear that the return will definitely be `Index`, could maybe add a code comment adjacent to to cast with the justification of the cast.

If doing that I would only add the cast for `lev` and leave `lab` to be inferred. But I would probably prefer to leave the ignore until the other issues are resolved.

(I think we still have much of the lower level code to type)

Comment (twoertwein) [reply pandas/core/groupby/generic.py:629]: Based on the signature of `algorithms.factorize` I would expect that `lev` is assumed to be `np.ndarray | Index`. I would assume that there is a relationship between the first argument and the type of the second return value (`lev`) but i'm not sure how easy it will be to write overloads for this.
https://github.com/pandas-dev/pandas/blob/1b3bb88008011f8ce7e3ec65fcc0c31fb838352b/pandas/core/algorithms.py#L581

Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:629]: yep, wherever we have union return types we probably need overloads eventually, but my point is that I don't think that adding the overloads alone would fix the issue, since if the argument is `Any`, we would still return the union.

(side note: I'm sure i've tackled factorize in one of my stashes in the past, but the problem in many cases like this is overlapping overloads. (in this case IIRC it's the categorical) We would need to add ignores (which is normally difficult to get through the review process) and to be sure that the overloads are actually working as intended by matching the correct overload need tests in place to test the overloads. This has been in the pipeline for quite a while and personally I have been trying to avoid adding overloads till the test framework is in place)

File: pandas/core/groupby/generic.py
Comment (simonjayhawkins) [inline pandas/core/groupby/generic.py:635]: I'm not a fan of variable annotations where they should not be necessary. When an expression resolves to `Any` as is the case here since the return of `cut` is untyped, the variable annotation behaves just like a cast and allows us to lie to mypy (if we wanted too or unintentially). How difficult is it to add the type annotations to `pandas.core.reshape.tile.cut`?

File: pandas/core/groupby/generic.py
Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:648]: A cast here is probably OK since mypy can't narrow from `is_interval_dtype`

I think I've look at using TypeGuard for some of the these functions, but unfortunately we allow `arr_or_dtype` as the argument so we would end up returning a union return type and so does not solve the issue in many cases. For TypeGuard to be effective we would have to limit the input to the is_* functions.

For clarity, I personally would put the cast directly adjacent to the `is_interval_dtype` condition. (in general I prefer the cast to be a single statement rather than inlining in other expressions)

File: pandas/core/groupby/generic.py
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:22]: we cant just use tuple?

File: pandas/core/groupby/generic.py
Comment (FactorizeD) [reply pandas/core/groupby/generic.py:629]: Judging by the discussion it seems that it is better to leave as is for now. I will remove it from the PR

File: pandas/core/groupby/generic.py
Comment (FactorizeD) [reply pandas/core/groupby/generic.py:635]: I will have a look on it on Tuesday and let you know what I find

File: pandas/core/groupby/generic.py
Comment (FactorizeD) [reply pandas/core/groupby/generic.py:22]: I tried to do that but was getting some error (unfortunately don't remember which one exactly). However, it seems that it will no longer be needed as it was only imported for line 629 but it will be left as it was in this PR

Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:22]: Yes, `tuple` can only be used in type annotations (function and variable). `cast` is actually a runtime construct that is a no-op but does need to be parsed by the interpreter nonetheless, so must import `Tuple` from typing in these cases.

> However, it seems that it will no longer be needed as it was only imported for line 629 but it will be left as it was in this PR

yep, just mentioned the above for clarification

File: pandas/core/groupby/generic.py
Comment (simonjayhawkins) [inline pandas/core/groupby/generic.py:94]: inside TYPE_CHECKING blocks can safely import from the top level without causing circular imports.

 ```suggestion
     from pandas import Interval
```

File: pandas/core/groupby/generic.py
Comment (FactorizeD) [reply pandas/core/groupby/generic.py:635]: @simonjayhawkins ok, so it seems that `pandas.core.reshape.tile.cut` could work but would require additional overloads. The return value has a lot of dependencies on cut's parameters. Should I try to fix it in a separate PR given one of your previous comments?
> I have been trying to avoid adding overloads till the test framework is in place

I would also like to try to get this PR to the merge'able state, should I simply leave this ignore here for the time being (and potentially fix it in the PR adding the overloads)?

Comment (simonjayhawkins) [reply pandas/core/groupby/generic.py:635]: > I would also like to try to get this PR to the merge'able state, should I simply leave this ignore here for the time being (and potentially fix it in the PR adding the overloads)?

sgtm.

because there is ongoing work to merge the MS Stubs into the pandas repository #45253 and add a test framework for the public api #45594, we need to be careful not to clash with that effort for now.

unfortunately #45594 which is IMO the most important aspect of the effort to merge the MS stubs has gone stale.

@FactorizeD I do appreciate the work done in this PR, but some of these ignores are better fixed after the MS Stubs are merged

The alternative maybe to open a precursor to this to bring in the types from the MS stubs for pandas.cut. I've not been following the progress of that and what the process is. @Dr-Irv?

Comment (Dr-Irv) [reply pandas/core/groupby/generic.py:635]: > The alternative maybe to open a precursor to this to bring in the types from the MS stubs for pandas.cut. I've not been following the progress of that and what the process is. @Dr-Irv?

The MS stubs were initially created by Microsoft by auto-generating them from pandas source.  At some point, the stub for `pd.cut()` had no type annotations and was in the repo, but has since disappeared.  Now, there is no stub at all, so that's something that will have to be added at some point.

Comment (FactorizeD) [reply pandas/core/groupby/generic.py:635]: Ok, in this case I will leave it for now

**Comment by FactorizeD** [comment]
After reviewing all the comments it seems that the only change that can be applied at the moment is `cast` from line 648. Given that in general it is better to omit such, would you prefer me to still go with this one change, or should I close this PR and come back to it after e.g. stubs are merged?

**Comment by simonjayhawkins** [comment]
Thanks @FactorizeD can you get this PR into a mergeable state (and not marked as draft) even if only the one change.

**Review by twoertwein — APPROVED** [review]

**Comment by jreback** [comment]
can you merge master and fix the precommit

**Review by simonjayhawkins — APPROVED** [review]
Thanks @FactorizeD ci failure for `TestParquetPyArrow.test_unsupported` unrelated (and now fixed on main)

**Pull request merged**
